File: python/medium/longestSubstringwithoutrepeatingCharacters.py
#Leetcode number: 3

#Given a string s, find the length of the longest substring without repeating characters.



class Solution(object):
    def findLast(self,s,c):
        pos=-1
        i=-1
        for ch in s:
            pos+=1
            if ch==c:
                i=pos
        return i
        
    def lengthOfLongestSubstring(self, s):
        """
        :type s: str
        :rtype: int
        """
        sl=['']
        for c in s:
            i=self.findLast(sl[-1],c)
            if i<0:
                sl[-1]+=c
            else:
                sl.append(sl[-1][i+1:]+c)
        return max(map(len,sl))

File: java/medium/longestSubstringwithoutrepeatingCharacters.java
//Leetcode number: 3

//Given a string s, find the length of the longest substring without repeating characters.

class Solution {
    public int lengthOfLongestSubstring(String s) {
        if (s.length() == 0) return 0;
        
        int[] count = new int[256];
        for (int i = 0; i < 256; i++) count[i] = -1;
        count[(int) s.charAt(0)] = 0;

        int start = 0;
        int max = 1;
        
        for (int i = 1; i < s.length(); i++) {
            char ch = s.charAt(i);
            if (count[(int) ch] != -1) {
                start = Math.max(start, count[(int) ch]+1); // Math.max because for abba input;
            } 
            
            max = Math.max(max, i-start+1);
            count[(int) ch] = i;
        }
        
        return max;
        
    }
}